WIP: Add a boundary tracing algorithm

fbidu opened the pull request:
## Description

This is a work in progress implementation of the boundary tracing algorithm that started at SciPy sprints. It is not ready for review yet

[Thread on Zulip](https://skimage.zulipchat.com/#narrow/stream/176555-sprint/topic/boundary.20tracing)

Might close #1131 

## Checklist

<!-- It's fine to submit PRs which are a work in progress! -->
<!-- But before they are merged, all PRs should provide: -->
- [Docstrings for all functions](https://github.com/numpy/numpy/blob/master/doc/example.py)
- Gallery example in `./doc/examples` (new features only)
- Benchmark in `./benchmarks`, if your changes aren't covered by an
  existing benchmark
- Unit tests
- Clean style in [the spirit of PEP8](https://www.python.org/dev/peps/pep-0008/)

<!-- For detailed information on these and other aspects see -->
<!-- the scikit-image contribution guidelines. -->
<!-- https://scikit-image.org/docs/dev/contribute.html -->

## For reviewers

<!-- Don't remove the checklist below. -->
- Check that the PR title is short, concise, and will make sense 1 year
  later.
- Check that new functions are imported in corresponding `__init__.py`.
- Check that new features, API changes, and deprecations are mentioned in
      `doc/release/release_dev.rst`.

fbidu committed:
Draft implementation of boundary tracing

fbidu committed:
Treat excpetion at moore neighbor and fix boundary tracing API

Boundary Tracing now accepts a coordinate instead of a Region
Property

fbidu committed:
Expose trace_boundary in the measure submodule

Also renames the function to better match the naming schema
used throughout the module

fbidu committed:
Add test for a basic case of boundary detecting

fbidu committed:
Add horse test for boundary tracing

pep8speaks commented:
Hello @fbidu! Thanks for opening this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:



* In the file [`skimage/measure/_boundary.py`](https://github.com/scikit-image/scikit-image/blob/de5fb49e0b8a94c5e6b258dac9317154a040198e/skimage/measure/_boundary.py):

> [Line 4:80](https://github.com/scikit-image/scikit-image/blob/de5fb49e0b8a94c5e6b258dac9317154a040198e/skimage/measure/_boundary.py#L4): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (132 > 79 characters)
> [Line 67:80](https://github.com/scikit-image/scikit-image/blob/de5fb49e0b8a94c5e6b258dac9317154a040198e/skimage/measure/_boundary.py#L67): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (86 > 79 characters)
> [Line 73:80](https://github.com/scikit-image/scikit-image/blob/de5fb49e0b8a94c5e6b258dac9317154a040198e/skimage/measure/_boundary.py#L73): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (87 > 79 characters)



* In the file [`skimage/measure/tests/test_boundary.py`](https://github.com/scikit-image/scikit-image/blob/de5fb49e0b8a94c5e6b258dac9317154a040198e/skimage/measure/tests/test_boundary.py):

> [Line 8:1](https://github.com/scikit-image/scikit-image/blob/de5fb49e0b8a94c5e6b258dac9317154a040198e/skimage/measure/tests/test_boundary.py#L8): [E302](https://duckduckgo.com/?q=pep8%20E302) expected 2 blank lines, found 1
> [Line 20:80](https://github.com/scikit-image/scikit-image/blob/de5fb49e0b8a94c5e6b258dac9317154a040198e/skimage/measure/tests/test_boundary.py#L20): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (80 > 79 characters)
> [Line 27:1](https://github.com/scikit-image/scikit-image/blob/de5fb49e0b8a94c5e6b258dac9317154a040198e/skimage/measure/tests/test_boundary.py#L27): [E302](https://duckduckgo.com/?q=pep8%20E302) expected 2 blank lines, found 1

emmanuelle commented on skimage/measure/_boundary.py:
```diff
@@ -0,0 +1,96 @@
+    Parameters
+    ----------
+    coords : obj
+        Obtained with skimage.measure.regionprops()
```
```suggestion
        Obtained with skimage.measure.regionprops()
        
```

emmanuelle replied:
I tried adding a newline here because there is a formatting problem which makes the Python36-x64 build fail on the Azure CI

emmanuelle commented on skimage/measure/_boundary.py:
```diff
@@ -0,0 +1,96 @@
+
+def trace_boundary(coords):
+    """Coordinates of the region's boundary. The region must not have isolated
+    points.
```
```suggestion
    points.
    
```

emmanuelle commented:
Hi @fbidu great to see that you've made a lot of progress! Please tell us if you need some feedback now. You'll need to add a gallery example too since this is a new feature, probably in `doc/examples/segmentation`.

fbidu committed:
Update skimage/measure/_boundary.py

Co-authored-by: Emmanuelle Gouillart <emma@plot.ly>

fbidu committed:
Update skimage/measure/_boundary.py

Co-authored-by: Emmanuelle Gouillart <emma@plot.ly>

fbidu commented:
Hi @emmanuelle, I've just commited your suggestions. Thanks!

Currently, I'm still working on a new implementation of that function. The current one was a nice start and allowed me to write some tests but it needs some work to make it better. I'll be sure to add the examples later on and let you all know when I have something reviewable. Right know I'm still working on grokking the current behavior

fbidu committed:
refactor to anotate my current findings

At this point I'm just breaking apart some pieces of the original
function to add my notes after analyzing the current behavior

fbidu committed:
Merge branch 'boundary-tracing' of github.com:fbidu/scikit-image into boundary-tracing

rfezzani commented on skimage/measure/_boundary.py:
```diff
@@ -0,0 +1,160 @@
+        that this method will begin its search at (20, 20), a point very
+        far from the boundary
+        """
+        return np.amax(coords, axis=0)
```
Is it worth it defining the `starting_point` function for a one line call?

lagru replied:
No not really. And it's something that should be solved differently in Cython where there is no need to check every element. We don't really need the position of a global maximum but only the edge of the object.

rfezzani commented on skimage/measure/_boundary.py:
```diff
@@ -0,0 +1,160 @@
+        """
+        return np.amax(coords, axis=0)
+
+    maxs = starting_point()
```
Moreover, it is called only once...

rfezzani commented on skimage/measure/_boundary.py:
```diff
@@ -0,0 +1,160 @@
+        
+        return backtrack_start, start
+
+    backtrack_start, start = find_boundary_start()
```
The same as for the `starting_point` function, the `find_boundary_start` is called only once...

rfezzani commented on skimage/measure/_boundary.py:
```diff
@@ -0,0 +1,160 @@
+        current = neighbors_ids[idx]
+        counter += 1
+
+        if np.all(current == start) and np.all(backtrack == backtrack_start):
```
Why not directly use this condition to stop the while loop instead of the `while True`?..

FirefoxMetzger replied:
I don't think so. This looks like a bottom controlled loop to me; moving it to the top would mean we never enter the body given the current initialization.

rfezzani replied:
You simply need to correctly initialize you variables to enter the loop :wink:

rfezzani commented on skimage/measure/_boundary.py:
```diff
@@ -0,0 +1,160 @@
+        # initilization
+        # starting point is the most upper left point
+        idx_start = 0
+        while True:  # asserting that the starting point is not isolated
```
```suggestion
        for start in zip(y, x):
```

rfezzani commented on skimage/measure/_boundary.py:
```diff
@@ -0,0 +1,160 @@
+
+    while True:
+        neighbors_ids = _moore_neighborhood(current, backtrack)
+        y = neighbors_ids[:, 0]
+        x = neighbors_ids[:, 1]
```
```suggestion
        y, x = neighbors_ids.T
```

rfezzani commented on skimage/measure/_boundary.py:
```diff
@@ -0,0 +1,160 @@
+        boundary.append(current)
+        backtrack = neighbors_ids[idx - 1]
+        current = neighbors_ids[idx]
+        counter += 1
```
I don't see `counter` used elsewhere...
```suggestion
```

rfezzani commented on skimage/measure/_boundary.py:
```diff
@@ -0,0 +1,160 @@
+        raise RuntimeError("The backtrack is not on the neighborhood")
+
+
+def trace_boundary(coords):
```
Since the function is not used for tracing, can we find an other name (`region_boundary`?)

lagru replied:
... or even `object_boundary`. "region" brings `regionprops` to mind while "object" fit better within the morphological context.

rfezzani reviewed (commented):
Thank you @fbidu, I left some comments. Is there any reference that we can cite to describe the used method?

FirefoxMetzger commented on skimage/measure/tests/boundary_fixtures.py:
```diff
@@ -0,0 +1,2060 @@
+import numpy as np
+
+HORSE_EXPECTED = np.array(
```
Why not save the array in a `.npz` file (compressed or uncompressed) using `np.save("expected_boundary_fixtures_horse.npy", HORSE_EXPECTED)`? This will create a much smaller file and much cleaner PR.

You can use it in your test as
```
HORSE_EXPECTED = np.load("expected_boundary_fixtures_horse.npy")
```

fbidu replied:
Oh, that's neat! I didn't know it was possible and I'll definitely use it. Thanks!

fbidu commented:
Hello everybody! Sorry for the delay, I didn't have much time in the last couple of weeks to work on this PR but I'll do more work today.

@rfezzani thanks for your review and you're right on the 'functions for single calls' and so on. What happened is that the code you're seeing right now was mostly copied from another project. The idea right now is to completely rewrite it in order to become more optimizable. Given the fact that I'm not very familiar with this particular algorithm, I did a careful line by line debugging of its behavior while taking notes. Since I knew it would be a few days before I could lay my hands in this code again, I did a small refactoring in order to isolate the main parts I identified and put my (paper-based) notes on the code so it wouldn't get lost.

Thank you all for the comments! Have a great weekend :smile:

fbidu committed:
Move Boundary Tests fixtures to the data directory

fbidu committed:
Add draft reimplementation of the original moore algorithm

fbidu committed:
Fix draft moore implementation

lagru commented:
@fbidu Great that you are still working on this! I haven't forgotten you nevertheless this PR dropped a bit from my radar. I'll try to catch up till Friday.

FirefoxMetzger commented on skimage/measure/_boundary.py:
```diff
@@ -0,0 +1,160 @@
+
+    def find_boundary_start():
+
+        # initilization
+        # starting point is the most upper left point
+        idx_start = 0
+        while True:  # asserting that the starting point is not isolated
+            start = [y[idx_start], x[idx_start]]
+
+            # Focus Start is a 3x3 matrix centered around start
+            focus_start = binary[start[0] - 1 : start[0] + 2, start[1] - 1 : start[1] + 2]
+
+            # If the current pixel is not the only one set to one in its
+            # 3x3 region, we stop
+            if np.sum(focus_start) > 1:
+                break
+            idx_start += 1
```
This seems a bit unnecessary. We require that the region doesn't contain isolated points, or - if you agree with my suggestion - that it is simply connected. In this case we are guaranteed to find at least one neighbor that is also part of the region.

An exception (edge case) is if the region only has a single pixel, in which case your current implementation will throw an error because it runs out of bounds (either because of `y[idx_start]` or further down when using `idx_start` which now is `idx_start=len(coords)`.

I would replace this with simply grabbing the coordinate with the lowest indices.

FirefoxMetzger commented on skimage/measure/_boundary.py:
```diff
@@ -0,0 +1,160 @@
+    current = start
+    backtrack = backtrack_start
+    boundary = []
+    counter = 0
```
```suggestion
```

Looks like counter isn't used anymore.

FirefoxMetzger commented on skimage/measure/_boundary.py:
```diff
@@ -0,0 +1,160 @@
+
+    while True:
+        neighbors_ids = _moore_neighborhood(current, backtrack)
+        y = neighbors_ids[:, 0]
+        x = neighbors_ids[:, 1]
+        neighbors = binary[tuple([y, x])]
```
You control `_moore_neighborhood`; wouldn't it make sense to return `neighbors` in a format you can directly use instead of having to first do conversion so that you can use it?

FirefoxMetzger commented on skimage/measure/_boundary.py:
```diff
@@ -0,0 +1,160 @@
+        if np.all(current == start) and np.all(backtrack == backtrack_start):
+            break
+
+    return np.array(boundary)
```
This is a quite opinion based comment: I would prefer the output would be of the same format as `np.nonzero()`. The same goes for `measures.regionprops().coords`, but I don't know what the members think about this (@lagru )?

FirefoxMetzger reviewed (commented):
Nice job so far! I think Moore's Boundary Tracing will be a nice addition to skimage. I left some comments on the current implementation.

lagru commented on skimage/measure/_boundary.py:
```diff
@@ -0,0 +1,160 @@
+import numpy as np
+
+"""
+Code copied from https://github.com/machine-shop/deepwings/blob/master/deepwings/method_features_extraction/image_processing.py#L185
```
It seems that this code was written by one of @stefanv's students so I expect that this won't be a problem. Nevertheless you're right, we should make sure that everyone involved is in agreement and happy. @theobdt is it alright if we reuse your code here? It would probably be easiest if you could license it with a BSD-compatible license or if you explicitly stated your consent here (in case we have it). :pray: 

See [this comment on zulip](https://skimage.zulipchat.com/#narrow/stream/176555-sprint/topic/boundary.20tracing/near/203599349) by @stefanv.

theobdt replied:
Hello @lagru, yes I personally agree with this code being reused, as long as @stefanv is also ok with it. We also worked on an initial [Cython version](https://github.com/theobdt/boundary_tracing), which might be relevant.

stefanv replied:
Good with me!

emmanuelle commented:
Hi @fbidu how can we help you with this PR? I'd love to get this feature merged :-).

stefanv commented:
Agreed that this would be a great addition! Please let us know if we can help you complete it.

fbidu commented:
Hi team! Oh, I forgot about this, sorry :( So I don't have much context here anymore but I can try to make some progress over the weekend :thinking: 

Any tips on how it would be best to move forward is appreciated :pray:

lagru commented:
Hey @fbidu, no worries. We all did. :sweat_smile: 

> Any tips on how it would be best to move forward is appreciated

@theobdt mentioned a initial Cython version in https://github.com/scikit-image/scikit-image/pull/4853#discussion_r480078503. That might be a good place to compare. Otherwise a good place to start would probably be resolving the merge conflicts. After that get the test's working again if that's not the case. Otherwise, it looks like there are a good number of open comments that should probably be addressed. Let us know if you need help with that.

lagru commented on skimage/data/_registry.py:
```diff
@@ -128,6 +128,7 @@
     "registration/tests/data/TransformedX75Y75.png": "a1e9ead5f8e4a0f604271e1f9c50e89baf53f068f1d19fab2876af4938e695ea",
     "data/cells.tif": "2120cfe08e0396324793a10a905c9bbcb64b117215eb63b2c24b643e1600c8c9",
     "data/mitosis.tif": "2751ba667c4067c5d30817cff004aa06f6f6287f1cdbb5b8c9c6a500308cb456",
+    "data/boundary_tracing_tests.npz": "1b37154113ac3964d43ef4f1fe93f7377e85a29750a84487f8e101be388b017f",
```
I think we try to limit / reduce the number of data files we have in the repository itself. Especially if they are highly specific to a few tests. In case we decide this is necessary https://gitlab.com/scikit-image/data is probably a better place for this file. :slightly_smiling_face:

stefanv replied:
Or re-use an existing image for testing.

matthew-brett commented:
We have changed the base branch for this PR from the `main` branch to the `pre-megamove-integration` branch.

## Summary - what this means for your PR

You can continue to work on your PR, now based on `pre-megamove-integration`.  You should update your branch to the current state of `pre-megamove-integration` (this is what the Github "Update branch" button does).   Please **do not merge the scikit-image `main` branch into your PR branch**.

Once we've merged your PR into the `pre-megamove-integration` branch, we'll update your changes for Megamove, and merge into the `main` branch.

In addition - our automated script detected that your branch had merge conflicts with the pre-Megamove branch. Please make sure to merge in the `pre-megamove-integration` branch, and resolve the merge conflicts in the way that you want.

Read on for the Megamove rationale, and more detailed instructions.

<details>

## Rationale

We've changed the base branch to make it easier and smoother to continue to work on your PR, while we work through a rather large change in the directory structure of the repository - a change that we have called *Megamove*.

See [the Megamove explanation issue](https://github.com/scikit-image/scikit-image/issues/8232) for more detail.

Changing the base branch means you do not see merge conflicts related to the *Megamove* change, and can continue to work with the repository in its original state, without worrying about porting those changes across to the new post-Megamove structure.

As you'll see in the linked issue, you can continue to work on your PR as before, and you can continue to use the Github GUI to "Update branch" - this updates from the new `pre-megamove-integration` branch.   However, there is one caveat: please, do not try and merge the `scikit-image` `main` branch into this one, but instead, keep track of the `pre-megamove-integration` branch.  The issue explains more.

To emphasize again, **do not try and merge the scikit-image `main` branch**.  You will get Megamove related merge conflicts that are rather tricky to resolve correctly.

When we've approved your PR, we'll merge into the `pre-megamove-integration` branch, and then process it so we can merge it into the `main` branch.   We'll let you know when we've merged your PR changes into `main`.

</details>

Any questions, please reply here.
